File: android/app/src/main/java/com/hivepulse/app/ui/apiaries/ApiaryViewModel.kt
package com.hivepulse.app.ui.apiaries

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hivepulse.app.data.api.ApiaryOut
import com.hivepulse.app.data.repository.ApiaryRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ApiaryListState(
    val apiaries: List<ApiaryOut> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null
)

@HiltViewModel
class ApiaryViewModel @Inject constructor(private val repo: ApiaryRepository) : ViewModel() {

    private val _state = MutableStateFlow(ApiaryListState())
    val state = _state.asStateFlow()

    init { load() }

    fun load() = viewModelScope.launch {
        _state.update { it.copy(isLoading = true, error = null) }
        runCatching { repo.list() }
            .onSuccess { list -> _state.update { it.copy(isLoading = false, apiaries = list) } }
            .onFailure { e  -> _state.update { it.copy(isLoading = false, error = e.message) } }
    }

    fun create(name: String, description: String?, lat: Double?, lon: Double?, address: String?, isPublic: Boolean, onDone: (String) -> Unit) =
        viewModelScope.launch {
            runCatching { repo.create(name, description, lat, lon, address, isPublic) }
                .onSuccess { a -> _state.update { it.copy(apiaries = it.apiaries + a) }; onDone(a.id) }
                .onFailure { e -> _state.update { it.copy(error = e.message) } }
        }

    fun update(id: String, name: String, description: String?, lat: Double?, lon: Double?, address: String?, isPublic: Boolean, onDone: () -> Unit) =
        viewModelScope.launch {
            runCatching { repo.update(id, name, description, lat, lon, address, isPublic) }
                .onSuccess { updated ->
                    _state.update { it.copy(apiaries = it.apiaries.map { a -> if (a.id == id) updated else a }) }
                    onDone()
                }
                .onFailure { e -> _state.update { it.copy(error = e.message) } }
        }

    fun delete(id: String) = viewModelScope.launch {
        runCatching { repo.delete(id) }
            .onSuccess { _state.update { it.copy(apiaries = it.apiaries.filter { a -> a.id != id }) } }
            .onFailure { e -> _state.update { it.copy(error = e.message) } }
    }

    fun clearError() = _state.update { it.copy(error = null) }
}

File: ios/HivePulse/ViewModels/ApiaryViewModel.swift
import Foundation

@MainActor
final class ApiaryViewModel: ObservableObject {
    @Published var apiaries: [ApiaryOut] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let service: any ApiaryServiceProtocol

    init(service: any ApiaryServiceProtocol = ApiaryService()) {
        self.service = service
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            let resp = try await service.list()
            apiaries = resp.items
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func create(name: String, description: String?, latitude: Double?, longitude: Double?, address: String?, isPublic: Bool = false) async throws -> ApiaryOut {
        let apiary = try await service.create(name: name, description: description, latitude: latitude, longitude: longitude, address: address, isPublic: isPublic)
        apiaries.append(apiary)
        return apiary
    }

    @discardableResult
    func update(_ id: String, name: String, description: String?, latitude: Double?, longitude: Double?, address: String?, isPublic: Bool? = nil) async throws -> ApiaryOut {
        let updated = try await service.update(id, name: name, description: description, latitude: latitude, longitude: longitude, address: address, isPublic: isPublic)
        if let idx = apiaries.firstIndex(where: { $0.id == id }) {
            apiaries[idx] = updated
        }
        return updated
    }

    func delete(_ id: String) async throws {
        try await service.delete(id)
        apiaries.removeAll { $0.id == id }
    }
}
